src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1642,40 +1645,69 @@ public override BoundNode VisitCall(BoundCall node)
         /// For each argument, figure out if its corresponding parameter is annotated with NotNullWhenFalse or
         /// EnsuresNotNull.
         /// </summary>
-        private static ImmutableArray<AttributeAnnotations> GetAnnotations(int numArguments,
+        private static ImmutableArray<FlowAnalysisAnnotations> GetAnnotations(int numArguments,
             bool expanded, ImmutableArray<ParameterSymbol> parameters, ImmutableArray<int> argsToParamsOpt)
         {
-            ArrayBuilder<AttributeAnnotations> builder = null;
+            ArrayBuilder<FlowAnalysisAnnotations> builder = null;
 
             for (int i = 0; i < numArguments; i++)
             {
                 (ParameterSymbol parameter, _) = GetCorrespondingParameter(i, parameters, argsToParamsOpt, expanded);
-                AttributeAnnotations annotations = parameter?.FlowAnalysisAnnotations ?? AttributeAnnotations.None;
+                FlowAnalysisAnnotations annotations = parameter?.FlowAnalysisAnnotations ?? FlowAnalysisAnnotations.None;
+
+                annotations = removeInapplicableAnnotations(parameter, annotations);
+
+                if (annotations != FlowAnalysisAnnotations.None && builder == null)
+                {
+                    builder = ArrayBuilder<FlowAnalysisAnnotations>.GetInstance(numArguments);
+                    builder.AddMany(FlowAnalysisAnnotations.None, i);
+                }
+
+                if (builder != null)
+                {
+                    builder.Add(annotations);
+                }
+            }
+
+            return builder == null ? default : builder.ToImmutableAndFree();
 
+            FlowAnalysisAnnotations removeInapplicableAnnotations(ParameterSymbol parameter, FlowAnalysisAnnotations annotations)
+            {
                 // Ignore NotNullWhenTrue that is inapplicable
                 annotations = removeInapplicableNotNullWhenSense(parameter, annotations, sense: true);
 
                 // Ignore NotNullWhenFalse that is inapplicable
                 annotations = removeInapplicableNotNullWhenSense(parameter, annotations, sense: false);
 
-                if (annotations != AttributeAnnotations.None && builder == null)
+                // AssertsTrue must be applied to a bool parameter
+                if ((annotations & FlowAnalysisAnnotations.AssertsTrue) != 0 &&
+                    (parameter.Type.SpecialType != SpecialType.System_Boolean))
```

Comment:
>parameter.Type.SpecialType != SpecialType.System_Boolean [](start = 21, length = 56)

Would it be simpler to check parameter's type and erase both AssertsTrue and AssertsFalse if the type is not boolean without checking the current state of flags?